engine/notifications.js

```diff
@@ -0,0 +1,283 @@
+      return sendConfirmation(endpoint, key, authSecret, deviceId, regFeatures)
+      .then(() => regFeatures);
+    }
+    return regFeatures;
+  });
+}
+
+// Registers to all notifications including new ones
+// Removes registrations to individual features
+function registerToAll(deviceId) {
+  return getRegisteredFeatures(deviceId)
+  .then(features => unregister(deviceId, features, true))
+  .then(() => Promise.all([
+    redis.sadd('all-notifications', deviceId),
+    redis.sadd(`${deviceId}-notifications`, 'all')]));
+}
+
+// registers to receive notifications
+// required info:
+// * deviceId
+// * list of feature slugs to register to
+//   special features are `new` and `all`
+//   * new - do notify about new features
+//   * all - all features including new, registering to all will remove
+//     all individual registrations
+// * endpoint [optional]
+// one can register to a feature without providing an endpoint
+// if it was already saved:
+// * key
+function register(deviceId, features, endpoint, key, authSecret) {
+  if (!features || features.length === 0) {
+    return Promise.reject(new Error('No features provided'));
+  }
+
+  return Promise.resolve()
+  .then(() => {
+    if (endpoint) {
+      const device = {
+        id: deviceId,
+        endpoint,
+        key: key || '',
+        authSecret: authSecret || '',
+      };
+      return redis.hmset(`device-${deviceId}`, device);
+    }
+  })
+  .then(() => redis.hgetall(`device-${deviceId}`))
+  .then(device => {
+    if (!device || !device.endpoint) {
+      throw new Error('No endpoint provided');
+    }
+  })
+  .then(() => {
+    if (features.indexOf('all') >= 0) {
+      return registerToAll(deviceId);
+    }
+    return Promise.all(
+      features.map(slug => [
+        redis.sadd(`${slug}-notifications`, deviceId),
+        redis.sadd(`${deviceId}-notifications`, slug)]));
+  })
+  .then(() => sendConfirmationToDevice(deviceId))
+  .then(() => getRegisteredFeatures(deviceId));
+}
+
+// update endpoint for the device
+// required:
+// * machineId
+// * endpoint
+function updateDevice(deviceId, endpoint, key, authSecret) {
+  if (!endpoint) {
+    return Promise.reject(new Error('No endpoint provided'));
+  }
+  return checkDeviceId(deviceId)
+  .then(() => redis.hmset(`device-${deviceId}`,
+                          'endpoint', endpoint, 'key', key, 'authSecret', authSecret));
+}
+
+function sendNotifications(feature, payload, isNew) {
+  return redis.smembers(`${feature}-notifications`)
+  .then(devices =>
+    redis.smembers('all-notifications')
+    .then(all => devices.concat(all))
+  )
+  .then(devices => {
+    if (isNew) {
+      return redis.smembers('new-notifications')
+      .then(fresh => devices.concat(fresh));
+    }
+    return devices;
+  })
+  .then(devices => Promise.all(
+        devices.map(deviceId => redis.hgetall(`device-${deviceId}`))))
+  .then(devices => Promise.all(
+    // XXX potential problem if many notifications to the same
```

Comment:
Do we have enough characters to show all of this? 
We might as a follow up show a message - 4 features updates and link to a "recently updated features" page.